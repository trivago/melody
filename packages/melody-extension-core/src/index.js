/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { unaryOperators, binaryOperators, tests } from './operators';
import { AutoescapeParser } from './parser/autoescape';
import { BlockParser } from './parser/block';
import { DoParser } from './parser/do';
import { EmbedParser } from './parser/embed';
import { ExtendsParser } from './parser/extends';
import { FilterParser } from './parser/filter';
import { FlushParser } from './parser/flush';
import { ForParser } from './parser/for';
import { FromParser } from './parser/from';
import { IfParser } from './parser/if';
import { ImportParser } from './parser/import';
import { IncludeParser } from './parser/include';
import { MacroParser } from './parser/macro';
import { SetParser } from './parser/set';
import { SpacelessParser } from './parser/spaceless';
import { UseParser } from './parser/use';
import { MountParser } from './parser/mount';

import forVisitor from './visitors/for';
import testVisitor from './visitors/tests';
import filters from './visitors/filters';
import functions from './visitors/functions';

const filterMap = [
    'attrs',
    'classes',
    'styles',
    'batch',
    'escape',
    'format',
    'merge',
    'nl2br',
    'number_format',
    'raw',
    'replace',
    'reverse',
    'round',
    'striptags',
    'title',
    'url_encode',
    'trim',
].reduce((map, filterName) => {
    map[filterName] = 'melody-runtime';
    return map;
}, Object.create(null));

Object.assign(filterMap, filters);

const functionMap = [
    'attribute',
    'constant',
    'cycle',
    'date',
    'max',
    'min',
    'random',
    'range',
    'source',
    'template_from_string',
].reduce((map, functionName) => {
    map[functionName] = 'melody-runtime';
    return map;
}, Object.create(null));
Object.assign(functionMap, functions);

export const extension = {
    tags: [
        AutoescapeParser,
        BlockParser,
        DoParser,
        EmbedParser,
        ExtendsParser,
        FilterParser,
        FlushParser,
        ForParser,
        FromParser,
        IfParser,
        ImportParser,
        IncludeParser,
        MacroParser,
        SetParser,
        SpacelessParser,
        UseParser,
        MountParser,
    ],
    unaryOperators,
    binaryOperators,
    tests,
    visitors: [forVisitor, testVisitor],
    filterMap,
    functionMap,
};

export {
    AutoescapeBlock,
    BlockStatement,
    BlockCallExpression,
    MountStatement,
    DoStatement,
    EmbedStatement,
    ExtendsStatement,
    FilterBlockStatement,
    FlushStatement,
    ForStatement,
    ImportDeclaration,
    FromStatement,
    IfStatement,
    IncludeStatement,
    MacroDeclarationStatement,
    VariableDeclarationStatement,
    SetStatement,
    SpacelessBlock,
    AliasExpression,
    UseStatement,
    UnaryNotExpression,
    UnaryNeqExpression,
    UnaryPosExpression,
    BinaryOrExpression,
    BinaryAndExpression,
    BitwiseOrExpression,
    BitwiseXorExpression,
    BitwiseAndExpression,
    BinaryEqualsExpression,
    BinaryNotEqualsExpression,
    BinaryLessThanExpression,
    BinaryGreaterThanExpression,
    BinaryLessThanOrEqualExpression,
    BinaryGreaterThanOrEqualExpression,
    BinaryNotInExpression,
    BinaryInExpression,
    BinaryMatchesExpression,
    BinaryStartsWithExpression,
    BinaryEndsWithExpression,
    BinaryRangeExpression,
    BinaryAddExpression,
    BinaryMulExpression,
    BinaryDivExpression,
    BinaryFloorDivExpression,
    BinaryModExpression,
    BinaryPowerExpression,
    BinaryNullCoalesceExpression,
    TestEvenExpression,
    TestOddExpression,
    TestDefinedExpression,
    TestSameAsExpression,
    TestNullExpression,
    TestDivisibleByExpression,
    TestConstantExpression,
    TestEmptyExpression,
    TestIterableExpression,
} from './types';
