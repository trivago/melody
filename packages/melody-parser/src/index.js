/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import Parser from './Parser';
import TokenStream from './TokenStream';
import * as Types from './TokenTypes';
import Lexer from './Lexer';
import { EOF, CharStream } from './CharStream';
import { LEFT, RIGHT } from './Associativity';
import {
    setStartFromToken,
    setEndFromToken,
    copyStart,
    copyEnd,
    copyLoc,
    getNodeSource,
    createNode,
    hasTagStartTokenTrimLeft,
    hasTagEndTokenTrimRight,
    isMelodyExtension,
} from './util';

function parse(code, options, ...extensions) {
    return createExtendedParser(code, options, ...extensions).parse();
}

function createExtendedParser(code, options, ...extensions) {
    let passedOptions = options;
    const passedExtensions = extensions;
    if (isMelodyExtension(options)) {
        // Variant without options parameter: createExtendedParser(code, ...extensions)
        passedOptions = undefined;
        passedExtensions.unshift(options);
    }
    const lexer = createExtendedLexer(code, options, ...passedExtensions);
    const parser = new Parser(
        new TokenStream(lexer, passedOptions),
        passedOptions
    );
    for (const ext of passedExtensions) {
        parser.applyExtension(ext);
    }
    return parser;
}

function createExtendedLexer(code, options, ...extensions) {
    let passedOptions = options;
    const passedExtensions = extensions;
    if (isMelodyExtension(options)) {
        // Variant without options parameter: createExtendedLexer(code, ...extensions)
        passedOptions = undefined;
        passedExtensions.unshift(options);
    }
    const lexer = new Lexer(new CharStream(code), passedOptions);
    for (const ext of passedExtensions) {
        lexer.applyExtension(ext);
    }
    return lexer;
}

export {
    Parser,
    TokenStream,
    Lexer,
    EOF,
    CharStream,
    LEFT,
    RIGHT,
    parse,
    createExtendedLexer,
    createExtendedParser,
    setStartFromToken,
    setEndFromToken,
    copyStart,
    copyEnd,
    copyLoc,
    getNodeSource,
    createNode,
    hasTagStartTokenTrimLeft,
    hasTagEndTokenTrimRight,
    Types,
};
