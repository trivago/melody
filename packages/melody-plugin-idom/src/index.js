/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as t from 'babel-types';
import mountVisitor from './visitors/mount.js';
import idomVisitor from './visitors/idom.js';

export default {
    visitors: [idomVisitor, mountVisitor],
    filterMap: {
        raw(path) {
            if (path.parentPath.is('PrintStatement')) {
                path.replaceWithJS(
                    t.expressionStatement(
                        t.callExpression(
                            t.identifier(
                                path.state.addImportFrom('melody-idom', 'raw')
                            ),
                            [path.get('target').node]
                        )
                    )
                );
            } else {
                path.replaceWithJS(
                    t.callExpression(
                        t.identifier(
                            path.state.addImportFrom('melody-idom', 'rawString')
                        ),
                        [path.get('target').node]
                    )
                );
            }
        },
    },
};
