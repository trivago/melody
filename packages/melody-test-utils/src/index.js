import {
    component,
    patchOuter,
    patchInner,
    flush,
    clearQueue,
    getNodeData,
    setCurrentComponent,
} from 'melody-idom';

import { print, test } from 'melody-snapshot-serializer';

import prettyFormat from 'pretty-format';

import { fireEvent } from 'dom-testing-library';

const flatten = nestedArray =>
    nestedArray.reduce((acc, cur) => acc.concat(cur), []);
const map = transform => data => data.map(transform);
const flatMap = transform => data => flatten(map(transform)(data));
const maybeUnwrap = data => (data.length === 1 ? data[0] : data);

const selectNodes = predicate => el => {
    const walker = document.createTreeWalker(el);
    const matches = [];
    while (walker.nextNode()) {
        if (predicate(walker.currentNode)) {
            matches.push(walker.currentNode);
        }
    }
    return matches;
};

const drainQueue = () =>
    flush({
        didTimeout: false,
        timeRemaining() {
            return 10;
        },
    });

const hasComponentOfType = Type => node => {
    const inst = getNodeData(node).componentInstance;
    return !!inst && inst instanceof Type;
};

const assertNotEmpty = (selector, component) => {
    if (component.elements.length === 0) {
        throw new Error(`Couldn't find component with selector: ${selector}`);
    }
    return component;
};
const exists = x => !!x;

const compose = (f, g) => x => f(g(x));

export const getComponentAt = node => getNodeData(node).componentInstance;

const shallowRenderComponent = comp => {
    setCurrentComponent(comp);
    patchOuter(comp.el, comp.render.bind(comp), {});
    clearQueue();
};

const dispatchTo = action => comp => {
    if (comp && comp.dispatch) {
        comp.dispatch(action);
        shallowRenderComponent(comp);
    }
};

const getComponentProps = compose(
    comp => comp && comp.props,
    getComponentAt
);

const getProp = key =>
    compose(
        props => props[key],
        getComponentProps
    );

const getComponentState = compose(
    comp => comp && ((comp.getState && comp.getState()) || comp.state),
    getComponentAt
);

export class Wrapper {
    constructor(rootElement) {
        this.elements = [].concat(rootElement);
    }

    get isMelodyWrapper() {
        return true;
    }

    get length() {
        return this.elements.length;
    }

    empty() {
        return this.elements.length === 0;
    }

    tap(callback) {
        callback(this);
        return this;
    }

    map(transform) {
        return this.lift(this.elements.map(transform));
    }

    reduce(reducer, initial) {
        return this.elements.reduce(reducer, initial);
    }

    forEach(callback) {
        this.elements.forEach(callback);
        return this;
    }

    filter(selector) {
        return this.filterWhere(x => x.is(selector));
    }

    filterWhere(predicate) {
        return this.lift(
            this.elements
                .map(this.lift)
                .filter(predicate)
                .reduce((acc, wrapper) => acc.concat(wrapper), this.lift([]))
        );
    }

    concat(other) {
        return this.lift(this.elements.concat(other.elements));
    }

    chain(transform) {
        return this.lift(flatMap(transform)(this.elements));
    }

    props() {
        return maybeUnwrap(this.elements.map(getComponentProps).filter(exists));
    }

    prop(key) {
        return maybeUnwrap(this.elements.map(getProp(key)).filter(exists));
    }

    state() {
        return maybeUnwrap(this.elements.map(getComponentState).filter(exists));
    }

    first() {
        return this.at(0);
    }

    last() {
        return this.at(this.length - 1);
    }

    at(index) {
        return this.lift(this.elements[index]);
    }

    get(index) {
        return this.elements[index];
    }

    dispatch(action) {
        this.elements.forEach(
            compose(
                dispatchTo(action),
                getComponentAt
            )
        );
        return this;
    }

    shallow() {
        const forceRender = comp => {
            if (comp) {
                shallowRenderComponent(comp);
            }
        };
        this.elements.forEach(
            compose(
                forceRender,
                getComponentAt
            )
        );
        return this;
    }

    render() {
        const forceRender = comp => {
            if (comp) {
                setCurrentComponent(comp);
                patchOuter(comp.el, comp.render.bind(comp), {});
            }
        };
        this.elements.forEach(
            compose(
                forceRender,
                getComponentAt
            )
        );
        drainQueue();
        return this;
    }

    setProps(props) {
        const setProps = comp => {
            if (comp && comp.apply) {
                comp.apply(props);
                shallowRenderComponent(comp);
            }
        };
        this.elements.forEach(
            compose(
                setProps,
                getComponentAt
            )
        );
        return this;
    }

    find(selector) {
        if (typeof selector === 'string') {
            return assertNotEmpty(
                selector,
                this.chain(el => Array.from(el.querySelectorAll(selector)))
            );
        } else if (typeof selector === 'function') {
            return assertNotEmpty(
                selector.prototype.displayName,
                this.chain(selectNodes(hasComponentOfType(selector)))
            );
        } else {
            throw new Error(
                `Your selector is invalid: ${selector} \n Make sure you give CSS selector or Component as selector`
            );
        }
    }

    findWhere(predicate) {
        return this.chain(
            selectNodes(
                compose(
                    predicate,
                    this.lift
                )
            )
        );
    }

    is(selector) {
        const test =
            typeof selector === 'string'
                ? el => el.matches(selector)
                : hasComponentOfType(selector);
        return this.length === 0
            ? false
            : this.reduce((acc, el) => acc && test(el), true);
    }

    hasClass(className) {
        return this.length === 0
            ? false
            : this.reduce(
                  (acc, el) => acc && el.classList.contains(className),
                  true
              );
    }

    html() {
        return this.reduce((acc, el) => acc + el.outerHTML, '');
    }

    text() {
        return this.reduce((acc, el) => acc + el.textContent, '');
    }

    simulate(type, options = { bubbles: true }) {
        const dispatchEvent = fireEvent[type];

        if (!dispatchEvent) {
            throw new TypeError(`No event found of type: "${type}"`);
        }

        this.forEach(el => dispatchEvent(el, options));
        drainQueue();
        return this;
    }

    debug() {
        return prettyFormat(this, {
            escapeRegex: true,
            plugins: [{ test, print }, prettyFormat.plugins.DOMElement],
            printFunctionName: false,
        });
    }

    // abstract lift(elements): Wrapper;
}

export class FullWrapper extends Wrapper {
    constructor(rootElement) {
        super(rootElement);
    }

    setProps(props) {
        const setProps = comp => {
            if (comp && comp.apply) {
                comp.apply(props);
                patchOuter(comp.el, comp.render.bind(comp), {});
            }
        };
        this.elements.forEach(
            compose(
                setProps,
                getComponentAt
            )
        );
        drainQueue();
        return this;
    }

    lift(rootElement) {
        return new FullWrapper(rootElement);
    }

    static empty() {
        return new FullWrapper([]);
    }
}

export class ShallowWrapper extends Wrapper {
    constructor(rootElement) {
        super(rootElement);
    }

    setProps(props) {
        const setProps = comp => {
            if (comp && comp.apply) {
                comp.apply(props);
                shallowRenderComponent(comp);
            }
        };
        this.elements.forEach(
            compose(
                setProps,
                getComponentAt
            )
        );
        return this;
    }

    lift(rootElement) {
        return new ShallowWrapper(rootElement);
    }

    static empty() {
        return new ShallowWrapper([]);
    }
}

export function render(Component, props = {}) {
    const el = document.createElement('div');
    patchInner(el, () => component(Component, 'key', props), {});
    drainQueue();
    return new FullWrapper(el.firstChild);
}

export function shallow(Component, props = {}) {
    const el = document.createElement('div');
    getNodeData(el).key = Symbol();
    patchInner(el, () => component(Component, 'key', props), {});
    const comp = getComponentAt(el.firstElementChild);
    shallowRenderComponent(comp);

    comp.el = el.firstElementChild;
    getNodeData(comp.el).componentInstance = comp;

    return new ShallowWrapper(comp.el);
}
