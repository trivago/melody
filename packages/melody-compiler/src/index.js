/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { CharStream, Lexer, TokenStream, Parser } from 'melody-parser';
import { traverse, merge, Path, Scope } from 'melody-traverse';

import { Template, File } from './Template.js';
import analyse from './analyse/index.js';
import convert from './convert/index.js';
import State from './state/State.js';
import finalizer from './finalizer/index.js';

import * as t from 'babel-types';
import { CodeGenerator } from 'babel-generator';

// workaround for https://github.com/rollup/rollup/issues/430
import { TokenTypes } from 'melody-parser';
export { Template, File, TokenTypes };

export function compile(fileName: String, source: String, ...extensions) {
    const root = parseString(fileName, source, ...extensions);
    const template = new Template(root.expressions),
        file = new File(fileName, template),
        state = new State(file, source);
    let analyseVisitor = analyse,
        convertVisitor = convert;
    for (const ext of extensions) {
        if (ext.visitors) {
            if (Array.isArray(ext.visitors)) {
                for (const visitor of (ext.visitors: Array)) {
                    if (visitor.analyse) {
                        analyseVisitor = merge(analyseVisitor, visitor.analyse);
                    }
                    if (visitor.convert) {
                        convertVisitor = merge(convertVisitor, visitor.convert);
                    }
                }
            } else {
                if (ext.visitors.analyse) {
                    analyseVisitor = merge(
                        analyseVisitor,
                        ext.visitors.analyse
                    );
                }
                if (ext.visitors.convert) {
                    convertVisitor = merge(
                        convertVisitor,
                        ext.visitors.convert
                    );
                }
            }
        }
        if (ext.filterMap) {
            Object.assign(state.filterMap, ext.filterMap);
        }
        if (ext.functionMap) {
            Object.assign(state.functionMap, ext.functionMap);
        }
        if (ext.options) {
            Object.assign(state.options, ext.options);
        }
    }
    convertVisitor = merge(convertVisitor, finalizer);
    const scope = Scope.get(
        Path.get({
            container: file,
            key: 'template',
        })
    );
    traverse(file, analyseVisitor, scope, state);
    traverse(file, convertVisitor, scope, state);
    return t.file(file.template);
}

function parseString(fileName: string, source: string, ...extensions) {
    const lexer = new Lexer(new CharStream(source));
    for (const ext of extensions) {
        lexer.applyExtension(ext);
    }
    const parser = new Parser(new TokenStream(lexer));
    for (const ext of extensions) {
        parser.applyExtension(ext);
    }

    return parser.parse();
}

export function toString(jsAst, code) {
    const g = new CodeGenerator(jsAst, {}, code);
    // Babel sanitises strings to not contain complex characters
    // however we need them in order to be able to render complex strings
    g._generator.StringLiteral = function(node, parent) {
        var raw = this.getPossibleRaw(node);
        if (!this.format.minified && raw != null) {
            this.token(raw);
            return;
        }

        return this.token(JSON.stringify(node.value));
    };
    return g.generate();
}
