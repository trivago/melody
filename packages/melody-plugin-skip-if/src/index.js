/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as t from 'babel-types';
import { Node, type, alias, visitor } from 'melody-types';
import template from 'babel-template';
import { Types, setStartFromToken, setEndFromToken } from 'melody-parser';

export class SkipIfBlock extends Node {
    constructor(target, expressions) {
        super();
        this.target = target;
        this.expressions = expressions;
    }
}
type(SkipIfBlock, 'SkipIfBlock');
alias(SkipIfBlock, 'Block');
visitor(SkipIfBlock, 'expressions');

export const SkipIfParser = {
    name: 'skip',
    parse(parser, token) {
        const tokens = parser.tokens;

        tokens.expect(Types.SYMBOL, 'if');
        const condition = tokens.expect(Types.SYMBOL);
        tokens.expect(Types.TAG_END);

        const skipIfBlock = new SkipIfBlock(
            condition,
            parser.parse((tokenText, token, tokens) => {
                return !!(
                    token.type === Types.TAG_START &&
                    tokens.nextIf(Types.SYMBOL, 'endskip')
                );
            }).expressions
        );

        setStartFromToken(skipIfBlock, token);
        setEndFromToken(skipIfBlock, tokens.expect(Types.TAG_END));

        return skipIfBlock;
    },
};

const buildSkipIfDefined = template(`
if (customElements.get(ELEMENT_NAME) !== undefined) {
    SKIP
} else {
    BODY
}
`);

export default {
    tags: [SkipIfParser],
    visitors: [
        {
            analyse: {
                SkipIfBlock(path) {
                    if (path.node.target.text === 'client') {
                        path.skip();
                    }
                },
            },
            convert: {
                SkipIfBlock: {
                    enter(path) {
                        if (path.node.target.text === 'client') {
                            path.replaceWithJS(skip(this));
                        }
                    },
                    exit(path) {
                        const target = path.node.target.text;
                        if (target === 'server') {
                            path.replaceWithMultipleJS(
                                ...path.node.expressions
                            );
                        } else if (target === 'defined') {
                            const el = path.findParentPathOfType('Element');
                            const elementName = el.node.name;
                            if (elementName.indexOf('-') === -1) {
                                this.error(
                                    'skip if defined can only be used inside of a custom element',
                                    el.node.elementNameLoc,
                                    `Custom Element must contain a "-" within their name but "${elementName}" does not seem to contain one.
More information about custom elements can be found here: https://developer.mozilla.org/en-US/docs/Web/Web_Components/Custom_Elements`,
                                    elementName.length
                                );
                            }
                            path.replaceWithJS(
                                buildSkipIfDefined({
                                    ELEMENT_NAME: t.stringLiteral(elementName),
                                    SKIP: skip(this),
                                    BODY: path.node.expressions,
                                })
                            );
                        } else {
                            this.error(
                                'Unknown skip if target',
                                path.node.target.pos,
                                `The target of the skip must be either 'client', 'server' or 'defined' but was "${target}" instead.`,
                                target.length
                            );
                        }
                    },
                },
            },
        },
    ],
};

function skip(state) {
    return t.expressionStatement(
        t.callExpression(
            t.identifier(state.addImportFrom('melody-idom', 'skip')),
            []
        )
    );
}
