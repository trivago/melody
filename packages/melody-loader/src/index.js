/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { toString, compile } from 'melody-compiler';
import { extension as CoreExtension } from 'melody-extension-core';
import { getOptions } from 'loader-utils';
import { isString, isObject } from 'lodash';

module.exports = function loader(content) {
    this.cacheable();

    const loaderOptions = getOptions(this) || {
        plugins: [],
    };

    const args = [this.resourcePath, content, CoreExtension];
    if (loaderOptions.plugins) {
        for (const pluginName of loaderOptions.plugins) {
            if (isString(pluginName)) {
                try {
                    args.push(require('melody-plugin-' + pluginName));
                } catch (e) {
                    this.emitWarning(
                        'Could not find plugin ' +
                            pluginName +
                            '. Expected name to be melody-plugin-' +
                            pluginName
                    );
                }
            } else if (isObject(pluginName)) {
                args.push(pluginName);
            } else {
                this.emitWarning(
                    'Value passed as Melody plugin must be string or object. ' +
                        pluginName +
                        ' of type ' +
                        typeof pluginName +
                        ' was given'
                );
            }
        }
    }

    try {
        const result = toString(compile.apply(null, args), content);
        return result.code;
    } catch (e) {
        this.emitError(e);
        return (
            'import {text} from "melody-idom"; export default { render(options) { text("Could not load ' +
            this.resourcePath +
            '"); console.error("Could not load ' +
            this.resourcePath +
            '", ' +
            JSON.stringify(e.message) +
            '); } };'
        );
    }
};
