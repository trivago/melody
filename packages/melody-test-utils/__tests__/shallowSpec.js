import List from './__fixtures__/list.js';
import Item, { clickRef as refFn } from './__fixtures__/item.js';
import Child from './__fixtures__/child.js';
import { shallow } from '../src';
import { RECEIVE_PROPS } from 'melody-component';

describe('Shallow rendering', () => {
    const initialProps = {
        items: [{ id: 1, name: 'Foo' }, { id: 2, name: 'Bar' }],
    };

    it('should return serialized object from debug', () => {
        const result = shallow(List, initialProps);
        expect(result.debug()).toMatchSnapshot();
    });

    it('should only render the top-level component', () => {
        const result = shallow(List, initialProps);
        expect(result).toMatchSnapshot();
    });

    it('should find the child Components', () => {
        const result = shallow(List, initialProps);
        const x = result.find(Item);
        expect(x).toMatchSnapshot();
    });

    it("should throw when find selectors don't match anything", () => {
        const result = shallow(Item, initialProps);
        expect(() => {
            result.find('xxx');
        }).toThrow(/^Couldn't find component with selector: xxx/);
    });

    it("should throw when find selectors don't match anything", () => {
        const result = shallow(Item, initialProps);
        expect(() => {
            result.find(List);
        }).toThrow(/^Couldn't find component with selector: List/);
    });

    it('should throw when find selectors are not a string or function', () => {
        const result = shallow(Item, initialProps);
        expect(() => {
            result.find();
        }).toThrow(/^Your selector is invalid: /);
        expect(() => {
            result.find({});
        }).toThrow(/^Your selector is invalid: /);
        expect(() => {
            result.find(123);
        }).toThrow(/^Your selector is invalid: /);
    });

    it('should get the props of a single component', () => {
        const result = shallow(List, initialProps);
        const x = result.find(Item).first();
        expect(x.props()).toMatchSnapshot();
    });

    it('should get the state of a single component', () => {
        const result = shallow(List, initialProps);
        const x = result.find(Item).first();
        expect(x.state()).toMatchSnapshot();
    });

    it('should allow dispatching actions', () => {
        const result = shallow(List, initialProps);
        const newProps = {
            items: [...initialProps.items, { id: 3, name: 'Baz' }],
        };
        result.dispatch({ type: RECEIVE_PROPS, payload: newProps });
        const x = result.find(Item);
        expect(x).toMatchSnapshot();
    });

    it('should allow setting props and rerender shallowly', () => {
        const result = shallow(List, initialProps);
        const newProps = {
            items: [...initialProps.items, { id: 3, name: 'Baz' }],
        };
        result.setProps(newProps);
        expect(result).toMatchSnapshot();
    });

    it('should allow filtering wrappers', () => {
        const result = shallow(List, {
            items: [...initialProps.items, { id: 3, name: 'Baz' }],
        });
        expect(
            result.find(Item).filterWhere(item => item.prop('id') !== 1)
        ).toMatchSnapshot();
    });

    it('should allow finding with a predicate function', () => {
        const result = shallow(List, initialProps);
        expect(
            result.findWhere(x => x.is(Item) || x.is(Child))
        ).toMatchSnapshot();
    });

    it('should allow filtering with a selector function', () => {
        const result = shallow(List, initialProps);
        expect(
            result.findWhere(x => x.is(Item) || x.is(Child)).filter(Child)
        ).toMatchSnapshot();
    });

    it('should unfold child components when their props are set', () => {
        const result = shallow(List, initialProps);
        result
            .find(Item)
            .at(1)
            .setProps({
                id: 2,
                name: 'Hello World',
            });
        expect(result).toMatchSnapshot();
    });

    it('should allow unfolding shallowly rendered children', () => {
        const result = shallow(List, initialProps);
        result
            .find(Item)
            .at(1)
            .render();
        expect(result).toMatchSnapshot();
    });

    it('should allow unfolding shallowly rendered children completely', () => {
        const result = shallow(List, initialProps);
        result.find(Child).render();
        expect(result).toMatchSnapshot();
    });

    it('should allow unfolding shallowly rendered children shallowly', () => {
        const result = shallow(List, initialProps);
        result.find(Child).shallow();
        expect(result).toMatchSnapshot();
    });

    it('should return the first', () => {
        const result = shallow(List, initialProps);
        expect(
            result
                .find(Item)
                .first()
                .props().id
        ).toEqual(1);
    });

    it('should return the prop value', () => {
        const result = shallow(List, initialProps);
        expect(
            result
                .find(Item)
                .first()
                .prop('id')
        ).toEqual(1);
    });
});
