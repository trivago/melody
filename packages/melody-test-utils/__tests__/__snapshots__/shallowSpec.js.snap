// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shallow rendering should allow dispatching actions 1`] = `
ShallowWrapper {
  "elements": Array [
    <m-placeholder />,
    <m-placeholder />,
    <m-placeholder />,
  ],
}
`;

exports[`Shallow rendering should allow filtering with a selector function 1`] = `
ShallowWrapper {
  "elements": Array [
    ShallowWrapper {
      "elements": Array [
        <m-placeholder />,
      ],
    },
  ],
}
`;

exports[`Shallow rendering should allow filtering wrappers 1`] = `
ShallowWrapper {
  "elements": Array [
    ShallowWrapper {
      "elements": Array [
        <m-placeholder />,
        <m-placeholder />,
      ],
    },
  ],
}
`;

exports[`Shallow rendering should allow finding with a predicate function 1`] = `
ShallowWrapper {
  "elements": Array [
    <m-placeholder />,
    <m-placeholder />,
    <m-placeholder />,
  ],
}
`;

exports[`Shallow rendering should allow setting props and rerender shallowly 1`] = `
ShallowWrapper {
  "elements": Array [
    <div>
      <ul>
        <m-placeholder />
        <m-placeholder />
        <m-placeholder />
      </ul>
      <m-placeholder />
    </div>,
  ],
}
`;

exports[`Shallow rendering should allow unfolding shallowly rendered children 1`] = `
ShallowWrapper {
  "elements": Array [
    <div>
      <ul>
        <m-placeholder />
        <li>
          <h2>
            Bar
          </h2>
        </li>
      </ul>
      <m-placeholder />
    </div>,
  ],
}
`;

exports[`Shallow rendering should allow unfolding shallowly rendered children completely 1`] = `
ShallowWrapper {
  "elements": Array [
    <div>
      <ul>
        <m-placeholder />
        <m-placeholder />
      </ul>
      <section>
        <span>
          This is awesome!
        </span>
      </section>
    </div>,
  ],
}
`;

exports[`Shallow rendering should allow unfolding shallowly rendered children shallowly 1`] = `
ShallowWrapper {
  "elements": Array [
    <div>
      <ul>
        <m-placeholder />
        <m-placeholder />
      </ul>
      <section>
        <m-placeholder />
      </section>
    </div>,
  ],
}
`;

exports[`Shallow rendering should find the child Components 1`] = `
ShallowWrapper {
  "elements": Array [
    <m-placeholder />,
    <m-placeholder />,
  ],
}
`;

exports[`Shallow rendering should get the props of a single component 1`] = `
Object {
  "clickRef": [Function],
  "id": 1,
  "name": "Foo",
}
`;

exports[`Shallow rendering should get the state of a single component 1`] = `
Object {
  "clickRef": [Function],
  "clicked": false,
  "id": 1,
  "name": "Foo",
}
`;

exports[`Shallow rendering should only render the top-level component 1`] = `
ShallowWrapper {
  "elements": Array [
    <div>
      <ul>
        <m-placeholder />
        <m-placeholder />
      </ul>
      <m-placeholder />
    </div>,
  ],
}
`;

exports[`Shallow rendering should return serialized object from debug 1`] = `
"<div>
  <ul>
    <Item
      id=1
      name=\\"Foo\\"
      clickRef=[Function]
/>
    <Item
      id=2
      name=\\"Bar\\"
      clickRef=[Function]
/>
  </ul>
  <Child />
</div>"
`;

exports[`Shallow rendering should unfold child components when their props are set 1`] = `
ShallowWrapper {
  "elements": Array [
    <div>
      <ul>
        <m-placeholder />
        <li>
          <h2>
            Hello World
          </h2>
        </li>
      </ul>
      <m-placeholder />
    </div>,
  ],
}
`;
